File: my-solutions/0020-valid-parentheses/solution.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<char> st = {};

        for (char c : s) {
            if (c == '(' || c == '{' || c == '[') {st.push_back(c);}
            else if (c == ')') {
                if (st.empty() || st.back() != '(') {return false;}
                else {st.pop_back();}
            } else if (c == '}') {
                if (st.empty() || st.back() != '{') {return false;}
                else {st.pop_back();}
            } else if (c == ']') {
                if (st.empty() || st.back() != '[') {return false;}
                else {st.pop_back();}
            }
        }

        return st.empty();
    }
};

File: my-solutions/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stk = []

        for c in s:
            if c in ("(", "[", "{"):
                stk.append(c)
            elif len(stk) == 0:
                return False
            elif (c == ")" and stk[-1] == "(") or (c == "]" and stk[-1] == "[") or (c == "}" and stk[-1] == "{"):
                stk.pop()
            else:
                return False
        
        return len(stk) == 0
